### PR TITLE
feat: add cloud option to supported method for storage

### LINKED ISSUE #2015: Method dynamicWhere() was not found in reflection of class Illuminate\Database\Query\Builder
**Internal error: Method dynamicWhere() was not found in reflection of class Illuminate\Database\Query\Builder. while analysing file**

image

- Larastan Version: 2.9.8
- Laravel Version: 10.38.1

### Description

 **Internal error: Method dynamicWhere() was not found in reflection of class Illuminate\Database\Query\Builder. while analysing file**

### stack trace

```
## phar:///var/www/html/project/vendor/phpstan/phpstan/phpstan.phar/src/Reflection/ClassReflection.php(564)
     #0 /var/www/html/project/vendor/larastan/larastan/src/Methods/BuilderHelper.php(116): PHPStan\Reflection\ClassReflection->getNativeMethod('dynamicWhere')
     #1 /var/www/html/project/vendor/larastan/larastan/src/Methods/BuilderHelper.php(211): Larastan\Larastan\Methods\BuilderHelper->dynamicWhere('whereIn', Object(PHPStan\Type\Generic\GenericObjectType))
     #2 /var/www/html/project/vendor/larastan/larastan/src/Methods/EloquentBuilderForwardsCallsExtension.php(93):
     Larastan\Larastan\Methods\BuilderHelper->searchOnEloquentBuilder(Object(PHPStan\Reflection\ClassReflection), 'whereIn', Object(PHPStan\Type\ObjectType))
     #3 /var/www/html/project/vendor/larastan/larastan/src/Methods/EloquentBuilderForwardsCallsExtension.php(47):
     Larastan\Larastan\Methods\EloquentBuilderForwardsCallsExtension->findMethod(Object(PHPStan\Reflection\ClassReflection), 'whereIn')
     #4 /var/www/html/project/vendor/larastan/larastan/src/Methods/ModelForwardsCallsExtension.php(197):
     Larastan\Larastan\Methods\EloquentBuilderForwardsCallsExtension->hasMethod(Object(PHPStan\Reflection\ClassReflection), 'whereIn')
     #5 /var/www/html/project/vendor/larastan/larastan/src/Methods/ModelForwardsCallsExtension.php(53): Larastan\Larastan\Methods\ModelForwardsCallsExtension->findMethod(Object(PHPStan\Reflection\ClassReflection),
     'whereIn')
     #6 phar:///var/www/html/project/vendor/phpstan/phpstan/phpstan.phar/src/Reflection/ClassReflection.php(512):
     Larastan\Larastan\Methods\ModelForwardsCallsExtension->hasMethod(Object(PHPStan\Reflection\ClassReflection), 'whereIn')
     #7 phar:///var/www/html/project/vendor/phpstan/phpstan/phpstan.phar/src/Type/ObjectType.php(558): PHPStan\Reflection\ClassReflection->hasMethod('whereIn')
     #8 phar:///var/www/html/project/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php(2139): PHPStan\Type\ObjectType->hasMethod('whereIn')
     #9 phar:///var/www/html/project/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php(2031): PHPStan\Analyser\NodeScopeResolver->processExprNode(Object(PhpParser\Node\Stmt\Return_),
     Object(PhpParser\Node\Expr\StaticCall), Object(PHPStan\Analyser\MutatingScope), Object(Closure), Object(PHPStan\Analyser\ExpressionContext))
     #10 phar:///var/www/html/project/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php(693): PHPStan\Analyser\NodeScopeResolver->processExprNode(Object(PhpParser\Node\Stmt\Return_),
     Object(PhpParser\Node\Expr\MethodCall), Object(PHPStan\Analyser\MutatingScope), Object(Closure), Object(PHPStan\Analyser\ExpressionContext))
     #11 phar:///var/www/html/project/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php(458): PHPStan\Analyser\NodeScopeResolver->processStmtNode(Object(PhpParser\Node\Stmt\Return_),
     Object(PHPStan\Analyser\MutatingScope), Object(Closure), Object(PHPStan\Analyser\StatementContext))
     #12 phar:///var/www/html/project/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/MutatingScope.php(1169): PHPStan\Analyser\NodeScopeResolver->processStmtNodes(Object(PhpParser\Node\Expr\Closure), Array,
     Object(PHPStan\Analyser\MutatingScope), Object(Closure), Object(PHPStan\Analyser\StatementContext))
     #13 phar:///var/www/html/project/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/MutatingScope.php(599): PHPStan\Analyser\MutatingScope->resolveType('function (array...', Object(PhpParser\Node\Expr\Closure))
     #14 phar:///var/www/html/project/vendor/phpstan/phpstan/phpstan.phar/src/Rules/FunctionCallParametersCheck.php(118): PHPStan\Analyser\MutatingScope->getType(Object(PhpParser\Node\Expr\Closure))
     #15 phar:///var/www/html/project/vendor/phpstan/phpstan/phpstan.phar/src/Rules/Methods/CallMethodsRule.php(48):
     PHPStan\Rules\FunctionCallParametersCheck->check(Object(PHPStan\Reflection\ResolvedFunctionVariantWithOriginal), Object(PHPStan\Analyser\MutatingScope), false, Object(PhpParser\Node\Expr\MethodCall), Array,
     'method')
     #16 phar:///var/www/html/project/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/FileAnalyser.php(120): PHPStan\Rules\Methods\CallMethodsRule->processNode(Object(PhpParser\Node\Expr\MethodCall),
     Object(PHPStan\Analyser\MutatingScope))
     #17 phar:///var/www/html/project/vendor/phpstan/phpstan/phpstan.phar/src/Node/ClassStatementsGatherer.php(109): PHPStan\Analyser\FileAnalyser->PHPStan\Analyser\{closure}(Object(PhpParser\Node\Expr\MethodCall),
     Object(PHPStan\Analyser\MutatingScope))
     #18 phar:///var/www/html/project/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php(647): PHPStan\Node\ClassStatementsGatherer->__invoke(Object(PhpParser\Node\Expr\MethodCall),
     Object(PHPStan\Analyser\MutatingScope))
     #19 phar:///var/www/html/project/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php(3226):
     PHPStan\Analyser\NodeScopeResolver::PHPStan\Analyser\{closure}(Object(PhpParser\Node\Expr\MethodCall), Object(PHPStan\Analyser\MutatingScope))
     #20 phar:///var/www/html/project/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php(1812): PHPStan\Analyser\NodeScopeResolver->callNodeCallbackWithExpression(Object(Closure),
     Object(PhpParser\Node\Expr\MethodCall), Object(PHPStan\Analyser\MutatingScope), Object(PHPStan\Analyser\ExpressionContext))
     #21 phar:///var/www/html/project/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php(693): PHPStan\Analyser\NodeScopeResolver->processExprNode(Object(PhpParser\Node\Stmt\Return_),
     Object(PhpParser\Node\Expr\MethodCall), Object(PHPStan\Analyser\MutatingScope), Object(Closure), Object(PHPStan\Analyser\ExpressionContext))
     #22 phar:///var/www/html/project/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php(458): PHPStan\Analyser\NodeScopeResolver->processStmtNode(Object(PhpParser\Node\Stmt\Return_),
     Object(PHPStan\Analyser\MutatingScope), Object(Closure), Object(PHPStan\Analyser\StatementContext))
     #23 phar:///var/www/html/project/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php(672): PHPStan\Analyser\NodeScopeResolver->processStmtNodes(Object(PhpParser\Node\Stmt\ClassMethod), Array,
     Object(PHPStan\Analyser\MutatingScope), Object(Closure), Object(PHPStan\Analyser\StatementContext))
     #24 phar:///var/www/html/project/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php(458): PHPStan\Analyser\NodeScopeResolver->processStmtNode(Object(PhpParser\Node\Stmt\ClassMethod),
     Object(PHPStan\Analyser\MutatingScope), Object(PHPStan\Node\ClassStatementsGatherer), Object(PHPStan\Analyser\StatementContext))
     #25 phar:///var/www/html/project/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php(783): PHPStan\Analyser\NodeScopeResolver->processStmtNodes(Object(PhpParser\Node\Stmt\Class_), Array,
     Object(PHPStan\Analyser\MutatingScope), Object(PHPStan\Node\ClassStatementsGatherer), Object(PHPStan\Analyser\StatementContext))
     #26 phar:///var/www/html/project/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php(458): PHPStan\Analyser\NodeScopeResolver->processStmtNode(Object(PhpParser\Node\Stmt\Class_),
     Object(PHPStan\Analyser\MutatingScope), Object(Closure), Object(PHPStan\Analyser\StatementContext))
     #27 phar:///var/www/html/project/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php(753): PHPStan\Analyser\NodeScopeResolver->processStmtNodes(Object(PhpParser\Node\Stmt\Namespace_), Array,
     Object(PHPStan\Analyser\MutatingScope), Object(Closure), Object(PHPStan\Analyser\StatementContext))
     #28 phar:///var/www/html/project/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/NodeScopeResolver.php(423): PHPStan\Analyser\NodeScopeResolver->processStmtNode(Object(PhpParser\Node\Stmt\Namespace_),
     Object(PHPStan\Analyser\MutatingScope), Object(Closure), Object(PHPStan\Analyser\StatementContext))
     #29 phar:///var/www/html/project/vendor/phpstan/phpstan/phpstan.phar/src/Analyser/FileAnalyser.php(179): PHPStan\Analyser\NodeScopeResolver->processNodes(Array, Object(PHPStan\Analyser\MutatingScope),
     Object(Closure))
     #30 phar:///var/www/html/project/vendor/phpstan/phpstan/phpstan.phar/src/Command/WorkerCommand.php(139): PHPStan\Analyser\FileAnalyser->analyseFile('/var/www/html/p...', Array,
     Object(PHPStan\Rules\LazyRegistry), Object(PHPStan\Collectors\Registry), NULL)
     #31 phar:///var/www/html/project/vendor/phpstan/phpstan/phpstan.phar/vendor/evenement/evenement/src/EventEmitterTrait.php(111): PHPStan\Command\WorkerCommand::PHPStan\Command\{closure}(Array)
     #32 phar:///var/www/html/project/vendor/phpstan/phpstan/phpstan.phar/vendor/clue/ndjson-react/src/Decoder.php(117): _PHPStan_c4c026984\Evenement\EventEmitter->emit('data', Array)
     #33 phar:///var/www/html/project/vendor/phpstan/phpstan/phpstan.phar/vendor/evenement/evenement/src/EventEmitterTrait.php(111): _PHPStan_c4c026984\Clue\React\NDJson\Decoder->handleData(Array)
     #34 phar:///var/www/html/project/vendor/phpstan/phpstan/phpstan.phar/vendor/react/stream/src/Util.php(62): _PHPStan_c4c026984\Evenement\EventEmitter->emit('data', Array)
     #35 phar:///var/www/html/project/vendor/phpstan/phpstan/phpstan.phar/vendor/evenement/evenement/src/EventEmitterTrait.php(111):
     _PHPStan_c4c026984\React\Stream\Util::_PHPStan_c4c026984\React\Stream\{closure}('{"action":"anal...')
     #36 phar:///var/www/html/project/vendor/phpstan/phpstan/phpstan.phar/vendor/react/stream/src/DuplexResourceStream.php(168): _PHPStan_c4c026984\Evenement\EventEmitter->emit('data', Array)
     #37 phar:///var/www/html/project/vendor/phpstan/phpstan/phpstan.phar/vendor/react/event-loop/src/StreamSelectLoop.php(201): _PHPStan_c4c026984\React\Stream\DuplexResourceStream->handleData(Resource id #7005)
     #38 phar:///var/www/html/project/vendor/phpstan/phpstan/phpstan.phar/vendor/react/event-loop/src/StreamSelectLoop.php(173): _PHPStan_c4c026984\React\EventLoop\StreamSelectLoop->waitForStreamActivity(NULL)
     #39 phar:///var/www/html/project/vendor/phpstan/phpstan/phpstan.phar/src/Command/WorkerCommand.php(99): _PHPStan_c4c026984\React\EventLoop\StreamSelectLoop->run()
     #40 phar:///var/www/html/project/vendor/phpstan/phpstan/phpstan.phar/vendor/symfony/console/Command/Command.php(259):
     PHPStan\Command\WorkerCommand->execute(Object(_PHPStan_c4c026984\Symfony\Component\Console\Input\ArgvInput), Object(_PHPStan_c4c026984\Symfony\Component\Console\Output\ConsoleOutput))
     #41 phar:///var/www/html/project/vendor/phpstan/phpstan/phpstan.phar/vendor/symfony/console/Application.php(870):
     _PHPStan_c4c026984\Symfony\Component\Console\Command\Command->run(Object(_PHPStan_c4c026984\Symfony\Component\Console\Input\ArgvInput), Object(_PHPStan_c4c026984\Symfony\Component\Console\Output\ConsoleOutput))
     #42 phar:///var/www/html/project/vendor/phpstan/phpstan/phpstan.phar/vendor/symfony/console/Application.php(261):
     _PHPStan_c4c026984\Symfony\Component\Console\Application->doRunCommand(Object(PHPStan\Command\WorkerCommand), Object(_PHPStan_c4c026984\Symfony\Component\Console\Input\ArgvInput),
     Object(_PHPStan_c4c026984\Symfony\Component\Console\Output\ConsoleOutput))
     #43 phar:///var/www/html/project/vendor/phpstan/phpstan/phpstan.phar/vendor/symfony/console/Application.php(157):
     _PHPStan_c4c026984\Symfony\Component\Console\Application->doRun(Object(_PHPStan_c4c026984\Symfony\Component\Console\Input\ArgvInput), Object(_PHPStan_c4c026984\Symfony\Component\Console\Output\ConsoleOutput))
     #44 phar:///var/www/html/project/vendor/phpstan/phpstan/phpstan.phar/bin/phpstan(124): _PHPStan_c4c026984\Symfony\Component\Console\Application->run()
     #45 phar:///var/www/html/project/vendor/phpstan/phpstan/phpstan.phar/bin/phpstan(125): _PHPStan_c4c026984\{closure}()
     #46 /var/www/html/project/vendor/phpstan/phpstan/phpstan(8): require('phar:///var/www...')
     #47 /var/www/html/project/vendor/bin/phpstan(119): include('/var/www/html/p...')
     #48 {main}
 -- ------------------
```

### PR DESCRIPTION
**Pull request by AhmedHelalAhmed**
- [x] Added or updated tests
- [ ] Documented user facing changes

**Changes**

Enhancement that add cloud option to storage supported methods

**Breaking changes**

### CODE REVIEW

**Comment by szepeviktor** [comment]
@AhmedHelalAhmed Thank you.
Could you add a test please.

**Review by szepeviktor — APPROVED** [review]
https://github.com/laravel/framework/blob/b5e1a9ff30921e4fa876ba05ab61ec76ed6a8797/src/Illuminate/Filesystem/FilesystemManager.php#L62-L96

File: src/ReturnTypes/StorageDynamicStaticMethodReturnTypeExtension.php [hunk 1]
```diff
@@ -25,7 +25,7 @@ public function getClass(): string
 
     public function isStaticMethodSupported(MethodReflection $methodReflection): bool
     {
-        return $methodReflection->getName() === 'disk' || $methodReflection->getName() === 'drive';
+        return in_array($methodReflection->getName(), ['disk', 'drive', 'cloud']);
```
Comment (szepeviktor) [inline src/ReturnTypes/StorageDynamicStaticMethodReturnTypeExtension.php]:
```suggestion
        return in_array($methodReflection->getName(), ['disk', 'drive', 'cloud'], strict: true);
```

and we could import it: `use function in_array;`

File: src/ReturnTypes/StorageDynamicStaticMethodReturnTypeExtension.php [hunk 2]
```diff
@@ -13,6 +13,7 @@
 use PHPStan\Type\DynamicStaticMethodReturnTypeExtension;
 use PHPStan\Type\ObjectType;
 use PHPStan\Type\Type;
+use function in_array;
```
Comment (calebdw) [inline src/ReturnTypes/StorageDynamicStaticMethodReturnTypeExtension.php:17]: code sniffer will fail, this needs a newline

Comment (AhmedHelalAhmed) [reply src/ReturnTypes/StorageDynamicStaticMethodReturnTypeExtension.php:16]: Ok

Comment (AhmedHelalAhmed) [reply src/ReturnTypes/StorageDynamicStaticMethodReturnTypeExtension.php:16]: It has a new line
image

Comment (AhmedHelalAhmed) [reply src/ReturnTypes/StorageDynamicStaticMethodReturnTypeExtension.php:16]: Can you clarify more, please?

Comment (calebdw) [reply src/ReturnTypes/StorageDynamicStaticMethodReturnTypeExtension.php:17]: The import groups (classes, functions, constants) need to be separated by a newline

```suggestion

use function in_array;
```

Comment (AhmedHelalAhmed) [reply src/ReturnTypes/StorageDynamicStaticMethodReturnTypeExtension.php:17]: I added space before it https://github.com/larastan/larastan/pull/2020/commits/d3abaaab1c931065f84649bf277b3b75be3b7bd1 
Thanks

**Pull request merged**